### ISSUE TITLE
Publish KDBX Binary Header Specification

### ISSUE THREAD

**Issue by seaofmars** [keepassxreboot/keepassxc-specs#8]
## Overview
Currently, there are no public written specifications of the KDBX 4 container format. There is an [RFC for its internal XML structure](https://github.com/keepassxreboot/keepassxc-specs/blob/master/kdbx-xml/rfc.txt), but not for its container format (i.e., the encryption layer).

## Expected Behavior
Developers wishing to re-implement, understand, or improve the KDBX spec, or developers fixing bugs in an existing software's implementation, may visit a repository of one of the major KeePass implementations, which would contain a description of the KDBX specification, not dissimilar from its internal format's RFC.

## Actual Behavior
Developers must reverse-engineer often ageing implementations from software to understand the format, and/or simply assume that what *other* implementations do is intended behavior.

There are repositories and gists providing a specification for some versions of KDBX, but none for KDBX 4 (that I found). Additionally, from my own efforts and from reading the KeePass source code, many sources are wrong.

I've previously posted on KeePass' help forum about this, to no avail.

**Comment by droidmonkey** [comment]
Good reference: https://palant.info/2023/03/29/documenting-keepass-kdbx4-file-format/#creating-a-composite-key

**Comment by jorabin** [comment]
Is your proposal to create such a reference here? 

I have written some documentation for KDBX, including version 4, but like you say it's not definitive. Well, nothing could be definitive without being endorsed by the original author. As you say, it's down to reading the source files provided by KeePass or reading some other implementation, here's mine in Java: https://github.com/jorabin/KeePassJava2/blob/master/kdbx/src/main/java/org/linguafranca/pwdb/kdbx/KdbxSerializer.java

@droidmonkey makes reference in their comment to some good documentation - see my comments there:

https://palant.info/2023/03/29/documenting-keepass-kdbx4-file-format/#c000003
https://palant.info/2023/03/29/documenting-keepass-kdbx4-file-format/#c000004

**Comment by droidmonkey** [comment]
Yes I think it is important that we add the binary header reference alongside our existing XML schema documentation. This is especially important for deciding where to take the KDBX standard in the future (see https://github.com/keepassxreboot/keepassxc/discussions/6229). KDBX is a relic of its time and there are very good reasons to move away from it, not least of which include how attachments and icons are handled, but also templates, merging custom data, shared groups. All these things are bolted together over the years of plugins and such.

We are staring at potentially implementing FIDO2 auth that abuses the current public custom data fields to make that work. The standard needs to come up to the times or be left behind.

**Comment by jorabin** [comment]
So, there is "official" XSD - https://keepass.info/help/download/KDBX_XML.xsd - and there is the the one here in this repo, and I have one too - https://github.com/jorabin/KeePassJava2/blob/master/KDBX.4.xsd - mine was created in the way it was to help with JAXB bindings (which I don't want any more, but ...). Illustrative of the fact that it's unlikely that there there would be one XSD to rule them all.

When you say: "The standard needs to come up to the times or be left behind." well, it's not a standard, there is no standardisation process, what there is is the code that Dominik continues to write in support of Windows KeePass. Some of us think that it would benefit from standardisation - i.e. a formally constituted organisation that documents and progresses the "standard", but until Dominik thinks that, KeePass is quite simply what he decides to make in his product. The things we write either track those changes, or not. If we don't track the changes we are not keeping up with formal compatibility (as defined by "it works in his version so it should work in mine and what works in mine needs to work in his").

Those of us who make compatible versions might want to define something new, but they would still need to track formal KeePass releases wouldn't they?

**Comment by droidmonkey** [comment]
Sure, in a strict sense it is not a standard, but having many implementations of it de facto makes it a standard to some extent. Dominik currently "controls" the features in the KDBX specification, but many projects, including ours, have defined "extensions" using custom data or advanced attributes. It is a really bad way of managing these features. Passkeys is a great example here.

If an actual KDBX standard were to be developed, KeePass2 would have to adopt it to remain compatible... not the other way around. We really should move away from KeePass2 setting the pace, its hold everything else back.

**Comment by phoerious** [comment]
Also, a standardisation process need not be democratic. If one person says so and everybody follows, it's a standard. If there's also a formal document for it, all the better.

**Comment by clach04** [comment]
@droidmonkey  Is https://keepass.info/help/kb/kdbx.html helpful?

 > KDBX File Format Specification
> Definition of the KDBX 4.1 file format.

Section https://keepass.info/help/kb/kdbx.html#header

**Comment by droidmonkey** [comment]
Yes that is the authoritative documentation.
